dotnet/src/webdriver/Remote/HttpCommandExecutor.cs

```diff
@@ -395,5 +392,57 @@ private class HttpResponseInfo
             public string Body { get; set; }
             public string ContentType { get; set; }
         }
+
+        /// <summary>
+        /// Internal diagnostic handler to log http requests/responses.
+        /// </summary>
+        private class DiagnosticsHttpHandler : DelegatingHandler
+        {
+            private readonly ILogger _logger;
+
+            public DiagnosticsHttpHandler(HttpMessageHandler messageHandler, ILogger logger)
+                : base(messageHandler)
+            {
+                _logger = logger;
+            }
+
+            /// <summary>
+            /// Sends the specified request and returns the associated response.
+            /// </summary>
+            /// <param name="request">The request to be sent.</param>
+            /// <param name="cancellationToken">A CancellationToken object to allow for cancellation of the request.</param>
+            /// <returns>The http response message content.</returns>
+            protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                StringBuilder requestLogMessageBuilder = new();
+                requestLogMessageBuilder.AppendFormat(">> {0}", request);
+
+                if (request.Content != null)
+                {
+                    var requestContent = await request.Content.ReadAsStringAsync().ConfigureAwait(false);
+                    requestLogMessageBuilder.AppendFormat("{0}{1}", Environment.NewLine, requestContent);
+                }
+
+                _logger.Trace(requestLogMessageBuilder.ToString());
+
+                var response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
```

Comment:
Can we send the request and handle the logging of it in parallel? Don't think we need to await reading the content and logging it before then awaiting sending?

Reply:
Micro optimization, but OK. I will do:
- send a request
- put log message
- await response

Reply:
Resolved.

Reply:
If logging were done in parallel, there is a possibility of interleaving logs from different requests, making it harder to follow the sequence of events. But @nvborisenko already have a commit for it so yeah nvm.

Reply:
@nvborisenko Sorry I meant send the request (but don't await) and then read the request content etc. and log it. As the key part and the slow part of the method I think is sending the request so if we get that done first and then handle the logging. Maybe its not necessary, up to you.

Reply:
Exactly, I implemented that as you said: kick off sending request, do logging, await response.

```csharp
var responseTask = base.SendAsync(request, cancellationToken).ConfigureAwait(false); // kick off sending request

_logger.Trace(requestLogMessageBuilder.ToString()); // do logging

var response = await responseTask; // await response
```

Reply:
@nvborisenko not quite, I said send the request and then read the request content and log it etc. The code as it is now has already read the request before it sends it. I would think that sending the request and awaiting response takes more time than reading the request so I thought it best to send first and then read it and log it while we are waiting.

Reply:
Ah, another micro optimization. Sorry, didn't catch it initially. I resolved it right now in `trunk`, please review.

Reply:
In any case your suggestions are valid, I am happy to implement it if they don't make code base fat.

Reply:
Thanks, no worries. I can't see any PRs open about it, maybe I am looking wrong. You can include a link perhaps?

Reply:
There is no PRs, I did it directly in trunk -  https://github.com/SeleniumHQ/selenium/commit/aba09e486d3c4f348a0a653d95b5bc94304c5587

Reply:
Looks decent, thanks. I eagerly await being able to investigate my issue further with this.